keps/1282-requeue-strategy/README.md

```diff
@@ -0,0 +1,306 @@
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [API Changes](#api-changes)
+  - [Changes to Queue Sorting](#changes-to-queue-sorting)
+    - [Existing Sorting](#existing-sorting)
+    - [Proposed Sorting](#proposed-sorting)
+    - [Preemption/Preemptor Flapping](#preemptionpreemptor-flapping)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Introduce new options that allow administrators to configure how Workloads are placed back in the queue after being after being evicted.
+
+## Motivation
+
+### Goals
+
+Allow administrators to configure requeuing behavior based on the reason the Workload is being requeued.
+
+### Non-Goals
+
+## Proposal
+
+Make queue placement after preemption configurable at the level of Kueue configuration.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Consider the following scenario:
+
+* A ClusterQueue has 2 ResourceFlavors.
+* Kueue has admitted a workload on ResourceFlavor #2.
+* There is a stock-out on the machine type needed to schedule this workload and the cluster autoscaler is unable to provision the necessary Nodes.
+* The workload gets evicted by Kueue because an administrator has configured the `waitForPodsReady` setting.
+* While the workload was pending capacity freed up on ResourceFlavor #1.
+
+In this case, the administrator would like the evicted workload to be requeued as soon as possible on the newly available capacity.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+
+## Design Details
+
+### API Changes
+
+Add additional fields to the Kueue ConfigMap to allow administrators to specify what timestamp to consider during queue sorting based on the reason for eviction.
+
+Possible settings:
+
+* `UseEvictionTimestamp` (Back of queue)
+* `UseCreationTimestamp` (Front of queue)
+
+```yaml
+kind: ConfigMap
+metadata:
+  name: kueue-manager-config
+  namespace: kueue-system
+data:
+  controller_manager_config.yaml: |
+    apiVersion: config.kueue.x-k8s.io/v1beta1
+    kind: Configuration
+    # ...
+    requeueStrategy:
+      priorityPreemption: UseEvictionTimestamp
+      podsReadyTimeout: UseCreationTimestamp
```

Comment:
We consider multiple aspects for sorting, the timestamp being one of them.

Having `UseCreationTimestamp` as the name of the option seems to imply that we ignore any other aspects.

Maybe it should be something like:
```
requeueStrategy:
  podsReadyTimeout:
    timestampSource: Creation | Eviction
```

That is assuming we are going to offer configuration for every eviction mechanism, otherwise, I would put the setting under the existing `waitForPodsReady`:

```
waitForPodsReady:
  requeuingTimestamp: Creation | Eviction
```

Reply:
I'm struggling to grasp how varying eviction methods could result in distinct types of queue sorting. Can you share some cases?